### ISSUE TITLE
Issues after upgrade to 2.0.4 (multi threading ?)

### ISSUE THREAD

**Issue by asteria277** [craigburke/angular-annotate-asset-pipeline#6]
Hi,

I'm trying to upgrade the grails plugin from 1.3.0 to 2.0.4. On my dev environment, I'm now having several errors when the javascript are loaded. For example : 

```
2015-02-10 14:03:44,467 [http-bio-8080-exec-6] ERROR [/JUMP_TRUNK].[default]  - Servlet.service() for servlet [default] in context with path [/JUMP_TRUNK] threw exception [Filter execution threw an exception] with root cause
Message: [error: couldn't process source due to parse error, Unexpected token (27:23)]
```

These errors are random. It's never the same lines numbers, or even the same files that fail.

So I've the impression there is a multi threading problem now. If two or more threads are sharing the same "globalScope", and if ng-annotate.js or Acorn are setting global variables, this could be the cause of those errors.

Note that version 2.0.3 doesn't seem to suffer from these problems (except when running under Jenkins... ?).

Just a suggestion, but in term of performance, would it not be better to maintain a cache ? 
If the file is not yet in the cache or has changed -> run ng-annotate, otherwise return the content of the cache.

**Comment by craigburke** [comment]
Hi @asteria277 that update was specifically to deal with multithreading, but I clearly missed something. I'll have to take another look.

The asset pipeline core should actually handle the caching.

**Comment by asteria277** [comment]
Thanks @craigburke !
Perhaps an @Synchronized('globalScope') should be added to the process method as well to make sure globalScope is not used by two threads at the same time ?
(and add the parameter 'globalScope' to the @Synchronized of initializeScope.

**Comment by craigburke** [comment]
I think you're exactly right :facepalm:

I'm going to move all the globalScope init stuff to a @Synchronized getter. I can't believe I overlooked that.

**Comment by craigburke** [comment]
@asteria277 could you see if you're still experiencing this with 2.0.5.

**Comment by asteria277** [comment]
Hi @craigburke . I'm working for another client today and don't have access to the project for testing. I will test 2.0.5 tomorrow morning and tell you if it works.

By the way, when I was looking at this yesterday I saw that ng-annotate was used to process all the javascript files. But most of them don't need to be processed. For example, jquery, jquery-ui, bootstrap, even angular.js ...
Do you know if there is a way to prevent this plugin to be used for all those files (in order to optimize the startup on a dev environment) ?

**Comment by craigburke** [comment]
@asteria277 awesome!

I believe the asset pipeline core supports includes and excludes config options for processors:
http://bertramdev.github.io/asset-pipeline/guide/configuration.html

I don't know if the syntax has changed since AP 2.0 but I'll take a look as I'm curious myself.

**Comment by asteria277** [comment]
Hi @craigburke . Unfortunately, I'm still having the same issues with 2.0.5...

I tried to make a patch but I don't know exactly how do build and test it since the grails plugin itself depends on another module.

But I can imagine that even with the new code, that globalScope can still be accessed at the same time by two different thread since the method "process" is not synchronized. So you can have two"ng-annotate.js" running at the same time, sharing the same global variables.

**Comment by craigburke** [comment]
@asteria277 bummer :frowning: 

I don't think the process method needs to be synchronized (I could be wrong). That's just creating an independent child scope based on the globalScope and calling the ngAnnotate function. 

Is there anyway you could provide one of the JS files that you see in your intermittent errors (emailing it to craig@craigburke.com). I just want to eliminate that as a possibility before potentially making the performance worse by synchronizing that method as well.

**Comment by asteria277** [comment]
Hi @craigburke . I agree that we should avoid having to make that method synchronized as this is not the best solution indeed.

I'm not a javascript expert, but I had the impression that ng-annotate was keeping a single instance of "acorn" in the global scope. If this is the case, and if two ngAnnotate() are executed at the same time, the input source could be mixed up while parsing a file resulting in the error I have. 

I will see tomorrow at work what I can send to you. Please note that my workstation is very slow, and I really have a lot of javascript files, so I've perhaps more chance to get the problem than someone else.

**Comment by craigburke** [comment]
@asteria277 you might very well be right. I'm not as familiar with the internals of ng-annotate as I should be. I might need to do some more investigation on that end.

Whatever you can send would be much appreciated. Thanks for working with me to figure this out!

**Comment by craigburke** [comment]
@asteria277 ok ok... I found another issue here. Apparently I was setting up my annotateScope incorrectly (it wasn't isolated like I thought it was). 

So give, 2.0.6 a try. Figures crossed this fixes it.

**Comment by asteria277** [comment]
@craigburke -> Unfortunatly I still got the same kind of errors with 2.0.6 :-(

I was able to reproduce the problem with the small script below. This script is using your class 'AnnotateProcessor'.

The first test is processing "jquery.js" 20 times sequentially. It never fails.
The second test is processing "jquery.js" in parallel (20 threads). It always fails at some point.

```java
import groovy.transform.Synchronized
import org.mozilla.javascript.Context
import org.mozilla.javascript.Scriptable

class AnnotateProcessor {

    static Scriptable globalScope

    AnnotateProcessor() { }

    @Synchronized()
    static Scriptable getGlobalScope() {
        if (!globalScope) {
            try {
                URL annotateResource = AnnotateProcessor.classLoader.getResource('ngannotate.js')
                Context context = Context.enter()
                globalScope = context.initStandardObjects()
                context.evaluateString(globalScope, annotateResource.text, annotateResource.file, 0, null)
            }
            catch (Exception ex) {
                throw new Exception("ngAnnotate initialization failed : " + ex.getMessage())
            }
            finally {
                try { Context.exit() }
                catch (Exception ex) {}
            }
        }

        globalScope
    }

    def process(String input) {
        try {
            def context = Context.enter()
            def annotateScope = context.newObject(AnnotateProcessor.globalScope)
            annotateScope.setPrototype(AnnotateProcessor.globalScope)
            annotateScope.setParentScope(null)
            annotateScope.put("inputSrc", annotateScope, input)

            Map result = (Map)context.evaluateString(annotateScope, "ngAnnotate(inputSrc, {add: true, sourcemap: false, stats: false})", "ngAnnotate command", 0, null)

            if (result.containsKey('errors')) {
                throw new Exception(result.errors as String)
            }
            else {
                return result.src
            }

        } catch (Exception ex) {
            throw new Exception("ngAnnotate failed: ${ex.message}", ex)
        } finally {
            try { Context.exit() }
            catch (Exception ex) {}
        }
    }

}


println "Starting ..."

URL jquery = AnnotateProcessor.classLoader.getResource('jquery-1.11.2.js')

println "Test 1 : no parallel processing"

(0..20).each { index ->
    println "Process : "+ index
    new AnnotateProcessor().process(jquery.text)
}

println "Test 2 : parallel processing"

def threads = (0..20).collect { index ->
    Thread.start {
        println "Starting thread : " + index
        new AnnotateProcessor().process(jquery.text)
    }
}

threads*.join()
```

**Comment by asteria277** [comment]
One possible solution could be to not create a shared global scope, but to just pre-compile the ng-annotate.js script with the following line in the synchronized method.
This requires more processing that using a global scope, but it more robust since we are sure everything is isolated ...
The following script works for me and doesn't give any errors anymore :

```javascript
import groovy.transform.Synchronized
import org.mozilla.javascript.Context
import org.mozilla.javascript.Script
import org.mozilla.javascript.Scriptable
import org.mozilla.javascript.ScriptableObject

class AnnotateProcessor {

    static Script annotateScript

    AnnotateProcessor() { }

    @Synchronized
    static Script getAnnotateScript() {
        if (! annotateScript) {
            try {
                URL annotateResource = AnnotateProcessor.classLoader.getResource('ngannotate.js')

                Context context = Context.enter()

                annotateScript = context.compileString(annotateResource.text, annotateResource.file, 0, null);
            }
            catch (Exception ex) {
                throw new Exception("ngAnnotate initialization failed : " + ex.getMessage())
            }
            finally {
                try { Context.exit() }
                catch (Exception ex) {}
            }
        }

        annotateScript
    }

    def process(String input) {
        try {
            def context = Context.enter()

            ScriptableObject annotateScope = context.initStandardObjects(null, true);
            getAnnotateScript().exec(context, annotateScope);

            annotateScope.put("inputSrc", annotateScope, input)

            Map result = (Map)context.evaluateString(annotateScope, "ngAnnotate(inputSrc, {add: true, sourcemap: false, stats: false})", "ngAnnotate command", 0, null)

            if (result.containsKey('errors')) {
                throw new Exception(result.errors as String)
            }
            else {
                return result.src
            }

        } catch (Exception ex) {
            throw new Exception("ngAnnotate failed: ${ex.message}", ex)
        } finally {
            try { Context.exit() }
            catch (Exception ex) {}
        }
    }

}


println "Starting ..."

URL jquery = AnnotateProcessor.classLoader.getResource('jquery-1.11.2.js')

def threads = (0..10).collect { index ->
    Thread.start {
        println "Starting thread : " + index
        new AnnotateProcessor().process(jquery.text)
    }
}

threads*.join()
```

**Comment by craigburke** [comment]
@asteria277 awesome awesome awesome! 

My only concern is that I know the initStandardObjects is really expensive. This is super helpful. I'll add that test to the current test suite so hopefully we can finally get this resolved.

Thanks so much!

**Comment by asteria277** [comment]
@craigburke yes it will be slower unfortunately ...
Another solution would be to keep a pool of 'globalScope' that could be reused by other threads.
When a thread needs a scope it would first try to get one available from the pool. If the pool is empty, then it would either wait for another thread to free one or create one. When the thread finishes it would put back the scope in the pool.
This would be more performant but this involve more coding...

**Comment by craigburke** [comment]
@asteria277 yeah, that's a good idea too. I really appreciate your insight on this.

**Comment by asteria277** [comment]
@craigburke You're welcome ! I use and like your plugins so I'm happy to contribute to it too !
Do you think you could release a new version so I can try it on monday ?
Thanks

**Comment by craigburke** [comment]
@asteria277 I'll certainly try to. Should have something either Monday or Tuesday.

**Comment by asteria277** [comment]
Hi @craigburke 

I wanted to test if the solution using a pool of 'globalScope' could improve the performance.

The answer is yes. On my (slow) machine, starting 10 threads running ng-annotate on jquery :

Without pool : 13 sec
With a pool : 9 sec

Note that the pool is limited to x instances (where x is the number of cores available on the machine). So the first x threads started will initialize the 'globalScope'. The other threads will wait until one of the thread is done and release a 'globalScope'.

```javascript
import groovy.transform.Synchronized
import groovy.util.logging.Log
import org.mozilla.javascript.Context
import org.mozilla.javascript.Script
import org.mozilla.javascript.ScriptableObject

class CacheItem {
    boolean inUse = false
    ScriptableObject script = null
    int pos
}

@Log
class AnnotateProcessor {

    static Script annotateScript

    AnnotateProcessor() { }

    private static final List<CacheItem> cache

    static {
        int cores = Runtime.getRuntime().availableProcessors();
        log.info("Number of cores : " + cores)

        cache = (1..cores).collect { new CacheItem(inUse: false, script: null, pos: it)}
    }

    @Synchronized
    static Script getAnnotateScript() {
        if (! annotateScript) {
            try {
                URL annotateResource = AnnotateProcessor.classLoader.getResource('ngannotate.js')

                Context context = Context.enter()

                annotateScript = context.compileString(annotateResource.text, annotateResource.file, 0, null);
            }
            catch (Exception ex) {
                throw new Exception("ngAnnotate initialization failed : " + ex.getMessage())
            }
            finally {
                try { Context.exit() }
                catch (Exception ex) {}
            }
        }

        annotateScript
    }

    static CacheItem getCacheItem() {
        synchronized (cache) {

            def free

            while (true) {
                free = cache.findAll { !it.inUse }
                if (! free.isEmpty()) {
                    def first = free.find { it.script } // Priority to item with a script already initialized
                    if (! first) {
                        first = free.first()
                    }
                    first.inUse = true
                    return first
                }

                log.info "No cache free ... wait"
                cache.wait()
            }
        }
    }

    static CacheItem getCacheItemAndCreateScript(Context context) {
        CacheItem cacheItem = getCacheItem()

        if (! cacheItem.script) {
            log.info "Init cache item : " + cacheItem.pos
            ScriptableObject annotateScope = context.initStandardObjects(null, true);
            getAnnotateScript().exec(context, annotateScope);

            cacheItem.script = annotateScope
        } else {
            log.info "Cache item already available : " + cacheItem.pos
        }

        return cacheItem
    }

    static void releaseCacheItem(CacheItem cacheItem) {
        synchronized (cache) {
            log.info "Release cache item : " + cacheItem.pos
            cacheItem.inUse = false
            cache.notify()
        }
    }

    def process(String input) {
        CacheItem cacheItem
        try {
            def context = Context.enter()

            cacheItem = getCacheItemAndCreateScript(context)

            cacheItem.script.put("inputSrc", cacheItem.script, input)
            Map result = (Map)context.evaluateString(cacheItem.script, "ngAnnotate(inputSrc, {add: true, sourcemap: false, stats: false})", "ngAnnotate command", 0, null)

            if (result.containsKey('errors')) {
                throw new Exception(result.errors as String)
            }
            else {
                return result.src
            }

        } catch (Exception ex) {
            throw new Exception("ngAnnotate failed: ${ex.message}", ex)
        } finally {
            try { Context.exit() }
            catch (Exception ex) {}
            if (cacheItem) {
                releaseCacheItem(cacheItem)
            }
        }
    }

}

Date now = new Date()
println "Starting ..."

URL jquery = AnnotateProcessor.classLoader.getResource('jquery-1.11.2.js')

def threads = (0..10).collect { index ->
    Thread.start {
        println "Starting thread : " + index
        new AnnotateProcessor().process(jquery.text)
    }
}

threads*.join()
println "TIME : " + (new Date().time - now.time)
```

**Comment by craigburke** [comment]
Whoa! Very impressive :open_mouth:  Thanks for working on this. 

I'd love to dig into this a little more later today. Regardless I'm thinking this is probably really close to what we're looking for. Why don't you create a PR for this (I want you to get credit for the code you wrote).

Thanks so much!

**Comment by asteria277** [comment]
Hi @craigburke . Yes I can try to create a PR for this. But I need to figure out how to test it since the code is not in the grails plugin itself. Could you please describe how you do it ?

**Comment by craigburke** [comment]
@asteria277 absolutely. You can pretty much ignore the grails plugin itself. It just adds the dependency for the core jar that is built from the project found in core (so just look at the project in the core folder).

There's a couple Spock tests there already, but really if you wanted to submit a PR with everything but the tests I can add them in after the fact.

**Comment by asteria277** [comment]
@craigburke I've created the pull request. I added a basic test and checked it fails without my fix. So there is no need to load a full script like jQuery, a simple line is enough.

**Comment by craigburke** [comment]
Thanks a lot! I'll take a look at this and I might hit you up with some questions but I'd love to roll your changes into a new release either this evening or tomorrow.

**Comment by asteria277** [comment]
Hi @craigburke . Don't want to put pressure, but are you still planning to release a new version ?
It's not very urgent, but I'm looking forward to upgrade my app with it :-)

**Comment by craigburke** [comment]
Hi @asteria277 yeah sorry, I had hoped to do that yesterday but I had to do some work on my document builder project to use at work.

By the way, do you use twitter? I'd love to give you a shout out there when I do release this.

**Comment by asteria277** [comment]
Hi @craigburke . No problem ! 
I'm not very active on twitter but I do have an account : erbr
I'm already following you.

**Comment by craigburke** [comment]
Your code looked great! I made some stylistic code changes (it helps me read and understand if I refactor), but I didn't make any substantive changes.

If you wouldn't mind taking a look at my latest commit to see if it all looks sane to you then I'll do a 2.1.0 release of this bad boy.

Thanks so much!

**Comment by asteria277** [comment]
Hi @craigburke . I checked the code and it looks good !
I will need to get the new release published to test with my app, but I don't see why it wouldn't work.
Thanks you very much too !

**Comment by craigburke** [comment]
Thanks for taking a look!

Go ahead and take version `2.1.0` for a test drive now if you'd like.

**Comment by asteria277** [comment]
2.1.0 is working perfectly !
Thanks again for this new version !

**Comment by craigburke** [comment]
WOOHOO! :tada: 

This version is all you, thanks again for the contribution!!

**Comment by craigburke** [comment]
@asteria277 I'd like to add you as a listed contributor. What name would you like me to use for that?

**Comment by asteria277** [comment]
Hi @craigburke . My name is "Eric Brayeur". I updated my github account with it.

**Comment by craigburke** [comment]
@asteria277 great! I added you to the contributors list. Thanks, man!
